File: 2022/day23/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use File::Spec;

# Parse elf positions from grid
sub parse_input {
    my ($text) = @_;
    my %elves;
    my @lines = split /\n/, $text;

    for my $r (0 .. $#lines) {
        my @chars = split //, $lines[$r];
        for my $c (0 .. $#chars) {
            if ($chars[$c] eq '#') {
                $elves{"$r,$c"} = 1;
            }
        }
    }
    return \%elves;
}

# Direction checks: check positions and move delta
my %dir_checks = (
    'N' => { checks => [[-1, -1], [-1, 0], [-1, 1]], delta => [-1, 0] },
    'S' => { checks => [[1, -1], [1, 0], [1, 1]], delta => [1, 0] },
    'W' => { checks => [[-1, -1], [0, -1], [1, -1]], delta => [0, -1] },
    'E' => { checks => [[-1, 1], [0, 1], [1, 1]], delta => [0, 1] },
);

# All 8 neighbors
my @all_neighbors = (
    [-1, -1], [-1, 0], [-1, 1],
    [0, -1], [0, 1],
    [1, -1], [1, 0], [1, 1]
);

# Run one round of simulation
sub simulate_round {
    my ($elves, $directions) = @_;

    my %proposals;       # elf_key -> proposed position key
    my %proposal_counts; # position key -> count

    # Phase 1: Each elf proposes a move
    for my $elf_key (keys %$elves) {
        my ($r, $c) = split /,/, $elf_key;

        # Check if any neighbors
        my $has_neighbor = 0;
        for my $n (@all_neighbors) {
            my $neighbor_key = ($r + $n->[0]) . "," . ($c + $n->[1]);
            if (exists $elves->{$neighbor_key}) {
                $has_neighbor = 1;
                last;
            }
        }

        next unless $has_neighbor;

        # Try each direction
        for my $d (@$directions) {
            my $checks = $dir_checks{$d}{checks};
            my $delta = $dir_checks{$d}{delta};

            my $can_move = 1;
            for my $check (@$checks) {
                my $check_key = ($r + $check->[0]) . "," . ($c + $check->[1]);
                if (exists $elves->{$check_key}) {
                    $can_move = 0;
                    last;
                }
            }

            if ($can_move) {
                my $new_pos = ($r + $delta->[0]) . "," . ($c + $delta->[1]);
                $proposals{$elf_key} = $new_pos;
                $proposal_counts{$new_pos}++;
                last;
            }
        }
    }

    # Phase 2: Execute moves (only if unique proposal)
    my %new_elves;
    my $moved = 0;

    for my $elf_key (keys %$elves) {
        if (exists $proposals{$elf_key}) {
            my $new_pos = $proposals{$elf_key};
            if ($proposal_counts{$new_pos} == 1) {
                $new_elves{$new_pos} = 1;
                $moved = 1;
            } else {
                $new_elves{$elf_key} = 1;
            }
        } else {
            $new_elves{$elf_key} = 1;
        }
    }

    return (\%new_elves, $moved);
}

# Count empty tiles in bounding rectangle
sub bounding_rect_empty {
    my ($elves) = @_;

    my ($min_r, $max_r, $min_c, $max_c);
    my $first = 1;

    for my $key (keys %$elves) {
        my ($r, $c) = split /,/, $key;
        if ($first) {
            $min_r = $max_r = $r;
            $min_c = $max_c = $c;
            $first = 0;
        } else {
            $min_r = $r if $r < $min_r;
            $max_r = $r if $r > $max_r;
            $min_c = $c if $c < $min_c;
            $max_c = $c if $c > $max_c;
        }
    }

    my $area = ($max_r - $min_r + 1) * ($max_c - $min_c + 1);
    return $area - scalar(keys %$elves);
}

# Part 1: Count empty tiles after 10 rounds
sub part1 {
    my ($text) = @_;
    my $elves = parse_input($text);
    my @directions = qw(N S W E);

    for (1 .. 10) {
        ($elves, my $moved) = simulate_round($elves, \@directions);
        push @directions, shift @directions;
    }

    return bounding_rect_empty($elves);
}

# Part 2: Find first round where no elf moves
sub part2 {
    my ($text) = @_;
    my $elves = parse_input($text);
    my @directions = qw(N S W E);

    my $round_num = 0;
    while (1) {
        $round_num++;
        my $moved;
        ($elves, $moved) = simulate_round($elves, \@directions);
        return $round_num unless $moved;
        push @directions, shift @directions;
    }
}

# Main
my $script_dir = dirname(File::Spec->rel2abs(__FILE__));
my $input_file = File::Spec->catfile($script_dir, '..', 'input.txt');

open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
my $text = do { local $/; <$fh> };
close $fh;

# Remove trailing newline
chomp $text;

print "Part 1: ", part1($text), "\n";
print "Part 2: ", part2($text), "\n";

File: 2022/day23/php/solution.php
#!/usr/bin/env php
<?php

function parseInput(string $text): array {
    $elves = [];
    $lines = explode("\n", trim($text));
    foreach ($lines as $r => $line) {
        for ($c = 0; $c < strlen($line); $c++) {
            if ($line[$c] === '#') {
                $elves["$r,$c"] = [$r, $c];
            }
        }
    }
    return $elves;
}

function simulateRound(array $elves, array $directions): array {
    // Direction checks: [check positions, move delta]
    $dirChecks = [
        'N' => [[[-1, -1], [-1, 0], [-1, 1]], [-1, 0]],
        'S' => [[[1, -1], [1, 0], [1, 1]], [1, 0]],
        'W' => [[[-1, -1], [0, -1], [1, -1]], [0, -1]],
        'E' => [[[-1, 1], [0, 1], [1, 1]], [0, 1]],
    ];

    // All 8 neighbors
    $allNeighbors = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]];

    // Phase 1: Each elf proposes a move
    $proposals = [];  // elf key -> proposed position key
    $proposalCounts = [];  // position key -> count

    foreach ($elves as $elfKey => $elf) {
        [$r, $c] = $elf;

        // Check if any neighbors
        $hasNeighbor = false;
        foreach ($allNeighbors as [$dr, $dc]) {
            $nKey = ($r + $dr) . "," . ($c + $dc);
            if (isset($elves[$nKey])) {
                $hasNeighbor = true;
                break;
            }
        }

        if (!$hasNeighbor) {
            continue;  // Don't move
        }

        // Try each direction
        foreach ($directions as $d) {
            [$checks, $delta] = $dirChecks[$d];
            [$dr, $dc] = $delta;

            $canMove = true;
            foreach ($checks as [$cr, $cc]) {
                $checkKey = ($r + $cr) . "," . ($c + $cc);
                if (isset($elves[$checkKey])) {
                    $canMove = false;
                    break;
                }
            }

            if ($canMove) {
                $newPos = ($r + $dr) . "," . ($c + $dc);
                $proposals[$elfKey] = $newPos;
                if (!isset($proposalCounts[$newPos])) {
                    $proposalCounts[$newPos] = 0;
                }
                $proposalCounts[$newPos]++;
                break;
            }
        }
    }

    // Phase 2: Execute moves (only if unique proposal)
    $newElves = [];
    $moved = false;

    foreach ($elves as $elfKey => $elf) {
        if (isset($proposals[$elfKey])) {
            $newPos = $proposals[$elfKey];
            if ($proposalCounts[$newPos] === 1) {
                $parts = explode(",", $newPos);
                $newElves[$newPos] = [(int)$parts[0], (int)$parts[1]];
                $moved = true;
            } else {
                $newElves[$elfKey] = $elf;
            }
        } else {
            $newElves[$elfKey] = $elf;
        }
    }

    return [$newElves, $moved];
}

function boundingRectEmpty(array $elves): int {
    $minR = PHP_INT_MAX;
    $maxR = PHP_INT_MIN;
    $minC = PHP_INT_MAX;
    $maxC = PHP_INT_MIN;

    foreach ($elves as [$r, $c]) {
        $minR = min($minR, $r);
        $maxR = max($maxR, $r);
        $minC = min($minC, $c);
        $maxC = max($maxC, $c);
    }

    $area = ($maxR - $minR + 1) * ($maxC - $minC + 1);
    return $area - count($elves);
}

function part1(string $text): int {
    $elves = parseInput($text);
    $directions = ['N', 'S', 'W', 'E'];

    for ($i = 0; $i < 10; $i++) {
        [$elves, $_] = simulateRound($elves, $directions);
        // Rotate directions
        $first = array_shift($directions);
        $directions[] = $first;
    }

    return boundingRectEmpty($elves);
}

function part2(string $text): int {
    $elves = parseInput($text);
    $directions = ['N', 'S', 'W', 'E'];

    $roundNum = 0;
    while (true) {
        $roundNum++;
        [$elves, $moved] = simulateRound($elves, $directions);
        if (!$moved) {
            return $roundNum;
        }
        // Rotate directions
        $first = array_shift($directions);
        $directions[] = $first;
    }
}

function main(): void {
    $scriptDir = dirname(__FILE__);
    $inputFile = $scriptDir . '/../input.txt';

    $text = file_get_contents($inputFile);

    echo "Part 1: " . part1($text) . "\n";
    echo "Part 2: " . part2($text) . "\n";
}

main();
